### PR TITLE
planner: fix RANGE COLUMNS partition prune gives wrong result with special collation

### LINKED ISSUE #57261: Partition Pruning gives wrong partition for RANGE COLUMNS with multiple columns and utf8mb4_0900_ai_ci collation
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```mysql
CREATE TABLE `t` (
 `a` varchar(255) COLLATE utf8mb4_0900_ai_ci NOT NULL,
 `b` varchar(255) COLLATE utf8mb4_0900_ai_ci NOT NULL
)
PARTITION BY RANGE COLUMNS(`a`,`b`)
(PARTITION `pLTi` VALUES LESS THAN ('i',''),
 PARTITION `pMax` VALUES LESS THAN (MAXVALUE,''));

insert into t values ("Q","Q");

-- Wrong partition, pLTi instead of pMax
explain select * from t where a like 'Q%';
select * from t where a like 'Q%';
-- Wrong partition, pLTi instead of pMax
explain select * from t where a = 'Q';
select * from t where a = 'Q';
-- OK, but with partition:all !
explain select * from t where a >= 'Q';
select * from t where a >= 'Q';

-- Wrong partition, pLTi instead of pMax
explain select * from t where a like 'q%';
select * from t where a like 'q%';
-- Wrong partition, pLTi instead of pMax
explain select * from t where a = 'q';
select * from t where a = 'q';
-- OK, but with partition:all !
explain select * from t where a >= 'q';
select * from t where a >= 'q';
```

### 2. What did you expect to see? (Required)
Use of partition `pMax` instead of `pLTi`

### 3. What did you see instead (Required)
```mysql
tidb> CREATE TABLE `t` (
    ->  `a` varchar(255) COLLATE utf8mb4_0900_ai_ci NOT NULL,
    ->  `b` varchar(255) COLLATE utf8mb4_0900_ai_ci NOT NULL
    -> )
    -> PARTITION BY RANGE COLUMNS(`a`,`b`)
    -> (PARTITION `pLTi` VALUES LESS THAN ('i',''),
    ->  PARTITION `pMax` VALUES LESS THAN (MAXVALUE,''));
Query OK, 0 rows affected (0.03 sec)

tidb> 
tidb> insert into t values ("Q","Q");
Query OK, 1 row affected (0.03 sec)

tidb> 
tidb> -- Wrong partition, pLTi instead of pMax
Query OK, 1 row affected (0.00 sec)

tidb> explain select * from t where a like 'Q%';
+-------------------------+----------+-----------+----------------+--------------------------------+
| id                      | estRows  | task      | access object  | operator info                  |
+-------------------------+----------+-----------+----------------+--------------------------------+
| TableReader_7           | 250.00   | root      | partition:pLTi | data:Selection_6               |
| └─Selection_6       | 250.00   | cop[tikv] |                | like(test.t.a, "Q%", 92)       |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t        | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+----------------+--------------------------------+
3 rows in set (0.00 sec)

tidb> select * from t where a like 'Q%';
Empty set (0.00 sec)

tidb> -- Wrong partition, pLTi instead of pMax
Query OK, 0 rows affected (0.00 sec)

tidb> explain select * from t where a = 'Q';
+-------------------------+----------+-----------+----------------+--------------------------------+
| id                      | estRows  | task      | access object  | operator info                  |
+-------------------------+----------+-----------+----------------+--------------------------------+
| TableReader_7           | 10.00    | root      | partition:pLTi | data:Selection_6               |
| └─Selection_6       | 10.00    | cop[tikv] |                | eq(test.t.a, "Q")              |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t        | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+----------------+--------------------------------+
3 rows in set (0.00 sec)

tidb> select * from t where a = 'Q';
Empty set (0.00 sec)

tidb> -- OK, but with partition:all !
Query OK, 0 rows affected (0.00 sec)

tidb> explain select * from t where a >= 'Q';
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 3333.33  | root      | partition:all | data:Selection_6               |
| └─Selection_6       | 3333.33  | cop[tikv] |               | ge(test.t.a, "Q")              |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
3 rows in set (0.00 sec)

tidb> select * from t where a >= 'Q';
+---+---+
| a | b |
+---+---+
| Q | Q |
+---+---+
1 row in set (0.00 sec)

tidb> 
tidb> -- Wrong partition, pLTi instead of pMax
Query OK, 0 rows affected (0.00 sec)

tidb> explain select * from t where a like 'q%';
+-------------------------+----------+-----------+----------------+--------------------------------+
| id                      | estRows  | task      | access object  | operator info                  |
+-------------------------+----------+-----------+----------------+--------------------------------+
| TableReader_7           | 250.00   | root      | partition:pLTi | data:Selection_6               |
| └─Selection_6       | 250.00   | cop[tikv] |                | like(test.t.a, "q%", 92)       |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t        | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+----------------+--------------------------------+
3 rows in set (0.00 sec)

tidb> select * from t where a like 'q%';
Empty set (0.00 sec)

tidb> -- Wrong partition, pLTi instead of pMax
Query OK, 0 rows affected (0.00 sec)

tidb> explain select * from t where a = 'q';
+-------------------------+----------+-----------+----------------+--------------------------------+
| id                      | estRows  | task      | access object  | operator info                  |
+-------------------------+----------+-----------+----------------+--------------------------------+
| TableReader_7           | 10.00    | root      | partition:pLTi | data:Selection_6               |
| └─Selection_6       | 10.00    | cop[tikv] |                | eq(test.t.a, "q")              |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t        | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+----------------+--------------------------------+
3 rows in set (0.00 sec)

tidb> select * from t where a = 'q';
Empty set (0.00 sec)

tidb> -- OK, but with partition:all !
Query OK, 0 rows affected (0.00 sec)

tidb> explain select * from t where a >= 'q';
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 3333.33  | root      | partition:all | data:Selection_6               |
| └─Selection_6       | 3333.33  | cop[tikv] |               | ge(test.t.a, "q")              |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
3 rows in set (0.00 sec)

tidb> select * from t where a >= 'q';
+---+---+
| a | b |
+---+---+
| Q | Q |
+---+---+
1 row in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v8.5.0-alpha-60-gb522e98080
Edition: Community
Git Commit Hash: b522e98080057cbaa451b207d42a8f37399082d7
Git Branch: HEAD
UTC Build Time: 2024-11-06 16:17:23
GoVersion: go1.23.2
Race Enabled: false
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #57261

Problem Summary: Don't use `sortKey` to do partition prune.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -1320,7 +1320,7 @@ func multiColumnRangeColumnsPruner(sctx base.PlanContext, exprs []expression.Exp
 		lens = append(lens, columnsPruner.partCols[i].RetType.GetFlen())
 	}
 
-	res, err := ranger.DetachCondAndBuildRangeForIndex(sctx.GetRangerCtx(), exprs, columnsPruner.partCols, lens, sctx.GetSessionVars().RangeMaxSize)
```
[the remark this answers is not in the document]
Comment (Defined2014) [reply pkg/planner/core/rule_partition_processor.go:1323]: Yeah, I will do it later. It's a draft.

[the remark this answers is not in the document]
Comment (Defined2014) [reply pkg/planner/core/rule_partition_processor.go:1323]: Done

File: pkg/planner/core/rule_partition_processor.go [hunk 2]
```diff
@@ -1335,16 +1335,12 @@ func multiColumnRangeColumnsPruner(sctx base.PlanContext, exprs []expression.Exp
 
 	rangeOr := make([]partitionRange, 0, len(res.Ranges))
 
-	comparer := make([]collate.Collator, 0, len(columnsPruner.partCols))
-	for i := range columnsPruner.partCols {
-		comparer = append(comparer, collate.GetCollator(columnsPruner.partCols[i].RetType.GetCollate()))
-	}
```
Comment (mjonss) [inline pkg/planner/core/rule_partition_processor.go:1341]: Is this just a simplification, since it was not needed, or does it also have impact on this PR?

**Review by mjonss — APPROVED** [review]
LGTM, with a minor question :)

File: pkg/planner/core/rule_partition_processor.go (hunk 2 above)
Comment (Defined2014) [reply pkg/planner/core/rule_partition_processor.go:1341]: just a simplification

**Review by qw4990 — APPROVED** [review]

**Pull request merged**